src/Umbraco.Cms.ManagementApi/Controllers/TreeControllerBase.cs

```diff
@@ -0,0 +1,119 @@
+﻿using Microsoft.AspNetCore.Mvc;
+using Umbraco.Cms.Core.Models;
+using Umbraco.Cms.Core.Models.Entities;
+using Umbraco.Cms.Core.Services;
+using Umbraco.Cms.ManagementApi.ViewModels.Tree;
+using Umbraco.Extensions;
+
+namespace Umbraco.Cms.ManagementApi.Controllers;
+
+public abstract class TreeControllerBase<TItem> : Controller
+    where TItem : TreeItemViewModel, new()
+{
+    private readonly string _itemUdiType;
+
+    protected TreeControllerBase(IEntityService entityService)
+    {
+        EntityService = entityService;
+
+        // ReSharper disable once VirtualMemberCallInConstructor
+        _itemUdiType = ItemObjectType.GetUdiType();
+    }
+
+    protected IEntityService EntityService { get; }
+
+    protected abstract UmbracoObjectTypes ItemObjectType { get; }
+
+    protected async Task<ActionResult<PagedResult<TItem>>> GetRoot(long pageNumber, int pageSize)
+    {
+        IEntitySlim[] rootEntities = GetPagedRootEntities(pageNumber, pageSize, out var totalItems);
+
+        TItem[] treeItemViewModels = MapTreeItemViewModels(null, rootEntities);
+
+        PagedResult<TItem> result = PagedResult(treeItemViewModels, pageNumber, pageSize, totalItems);
+        return await Task.FromResult(Ok(result));
+    }
+
+    protected async Task<ActionResult<PagedResult<TItem>>> GetChildren(Guid parentKey, long pageNumber, int pageSize)
+    {
+        IEntitySlim[] children = GetPagedChildEntities(parentKey, pageNumber, pageSize, out var totalItems);
+
+        TItem[] treeItemViewModels = MapTreeItemViewModels(parentKey, children);
+
+        PagedResult<TItem> result = PagedResult(treeItemViewModels, pageNumber, pageSize, totalItems);
+        return await Task.FromResult(Ok(result));
+    }
+
+    protected async Task<ActionResult<PagedResult<TItem>>> GetItems(Guid[] keys)
+    {
+        if (keys.IsCollectionEmpty())
+        {
+            return await Task.FromResult(Ok(PagedResult(Array.Empty<TItem>(), 0, 0, 0)));
+        }
+
+        IEntitySlim[] itemEntities = GetEntities(keys);
+
+        TItem[] treeItemViewModels = MapTreeItemViewModels(null, itemEntities);
+
+        var totalItems = itemEntities.Count();
+
+        PagedResult<TItem> result = PagedResult(treeItemViewModels, 0, totalItems, totalItems);
+        return await Task.FromResult(Ok(result));
+    }
+
+    protected virtual IEntitySlim[] GetPagedRootEntities(long pageNumber, int pageSize, out long totalItems)
+    {
+        // for now we'll just get all root entities; perhaps later on it will make sense to use EntityService.GetPagedChildren
```

Comment:
I think we should add a todo here, to make sure we circle back to it 😄

Reply:
It's not a TODO (yet) but will clarify the comment :)